lib/Mail/Milter/Authentication/Handler/SPF.pm

```diff
@@ -194,13 +198,153 @@ sub envfrom_callback {
         $self->log_error( 'SPF Error ' . $error );
         $self->{'failmode'} = 1;
         my $header = Mail::AuthenticationResults::Header::Entry->new()->set_key( 'spf' )->safe_set_value( 'temperror' );
-        $self->add_auth_header($header);
+        $self->{'spf_header'} = $header;
         $self->metric_count( 'spf_total', { 'result' => 'error' } );
     }
 }
 
+sub header_callback {
+    my ( $self, $header, $value ) = @_;
+
+    return unless exists $self->{'spfu_chain'};
+    return unless $self->{'dmarc_result'} eq 'pass';
+
+    my $lc_header = lc $header;
+
+    if ( $lc_header eq 'from') {
+        my $spfu_from_domain = lc $self->get_address_from($value);
+        $spfu_from_domain = $self->get_domain_from($spfu_from_domain) if $spfu_from_domain =~ /\@/;
+        $self->{'spfu_from_domain'} = $spfu_from_domain;
+
+        return;
+    }
+
+    if ( $lc_header eq 'received-spf' ||
+         $lc_header eq 'x-ms-exchange-authentication-results' ||
+         $lc_header eq 'arc-authentication-results' ||
+         $lc_header =~ 'authentication-results$'
+    ) {
+        push $self->{'spfu_chain'}->@*, { header => $header, value => $value };
+    }
+}
+
+sub eoh_callback {
+    my ($self) = @_;
+    return unless $self->{'spf_header'};
+    eval {
+        $self->spfu_checks();
+    };
+    $self->handle_exception( $@ );
+    $self->add_auth_header($self->{'spf_header'});
+}
+
+sub spfu_checks {
+    my ($self) = @_;
+
+    return unless exists $self->{'spfu_chain'};
+    return unless exists $self->{'spfu_from_domain'};
+    return unless $self->{'dmarc_result'} eq 'pass';
+    my $dmarc_object;
+    if ( $self->is_handler_loaded( 'DMARC' ) ) {
+        my $dmarc_handler = $self->get_handler('DMARC');
+        $dmarc_object = $dmarc_handler->get_dmarc_object();
+    }
+
+    my $spfu_from_domain = $self->{'spfu_from_domain'};
+    my $dmarc_domain = $self->{'dmarc_domain'};
+    if ($dmarc_object) {
+        # Work with org domain if possible
+        $dmarc_domain = $dmarc_object->get_organizational_domain( $dmarc_domain );
+        $spfu_from_domain = $dmarc_object->get_organizational_domain( $spfu_from_domain );
+    }
+
+    return unless lc $dmarc_domain eq $spfu_from_domain;
```

Comment:
For `$header` you `lc` to `$lc_header`.  So this adds a little confusion at this line whether `$spfu_from_domain` is lower cased.  

It is, it it created some confusion in my head